### ISSUE TITLE
Projections do not apply to discriminator properties as one would expect

### ISSUE THREAD

**Issue by jlchereau** [Automattic/mongoose#3133]
This is the base schema:
```
function VersionBaseSchema() {
    mongoose.Schema.apply(this, arguments);
    this.options.discriminatorKey = 'type';
    this.add({
        _summaryId          : { type: mongoose.Schema.Types.ObjectId, ref: 'Summary', required: true },
        _userId             : { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
        language            : { type: String, required: true, validate: schemaValidators.languageISOCode },
        state               : { type: Number, required: true, validate: schemaValidators.state },
        updated             : { type: Date, required: true, default: Date.now }
    });
}
var versionSchema = new VersionBaseSchema();
```

This is the discriminator:
```
var testSchema = new VersionBaseSchema({
    stream    : {
        pages           : { type: [pageSchema], default:[] },
        style           : { type: String, trim: true }
    }
});
model.discriminator('Test', testSchema);
```

This is the subdocument:
```
var pageSchema = new mongoose.Schema({
    components          : { type: [componentSchema], default:[] },
    style               : { type: String, trim: true }
});
```

The defect is illustrated by the following request:
```
var Version = mongoose.model('Version'),
    TypedVersion = Version.discriminators[version.type];

TypedVersion.findOne({ language: req.params.language, _summaryId: req.params.summaryId, _id: req.params.versionId }, { language: 1 }, function(error, version) {
         ...       
});
```

the returned version has:
- _id as expected,
- language as expected,

but also:
- type (or __t) which is unexpected, especially since {type: 0} removes the type from the projection;
- stream with its default value, {pages: []}, whatever the value in database, which is also unexpected.

**Comment by jlchereau** [comment]
Besides the discriminator type (or __t), which is a first issue, it seems that a projection does not apply to any property that is an object.

use a schema like:

```
var summarySchema = new mongoose.Schema({
    language: { type: String },
    complex: {
        prop1: { type: Number },
        prop2: { type: Number },
        prop3: { type: Number },
        prop4: { type: Number }
    }
});
```

and a projection on { language: 1} will also return complex with a value of {}.

**Comment by vkarpov15** [comment]
Sorry for taking so long to get back to you on this

Re: the '__t' inclusion, this behavior [is actually asserted on in the tests](https://github.com/Automattic/mongoose/blob/9f7d75bfe18b7b91e8d06bb15fe5e54a597c40ef/test/model.discriminator.querying.test.js#L516-L551). Mongoose relies on '__t' to figure out which schema to use when hydrating, and that's gonna be tricky to fix in addition to being a backwards breaking change. I added #3230 to track that.

Re: the second issue about inclusion on object types, that's unfortunately a known issue that's scheduled to be fixed in 5.0, see #2502. 

However, the 'stream.pages' issue getting its default value sounds like a separate bug. The below script doesn't repro that though:

```javascript
var mongoose = require('mongoose');
mongoose.set('debug', true);
mongoose.connect('mongodb://localhost/gh3133');

function VersionBaseSchema() {
    mongoose.Schema.apply(this, arguments);
    this.options.discriminatorKey = 'type';
    this.add({
        language            : { type: String, required: true },
        updated             : { type: Date, required: true, default: Date.now }
    });
}
VersionBaseSchema.prototype = mongoose.Schema.prototype;
var versionSchema = new VersionBaseSchema();

var componentSchema = new mongoose.Schema({ name: String });

var pageSchema = new mongoose.Schema({
    components          : { type: [componentSchema], default:[] },
    style               : { type: String, trim: true }
});

var testSchema = new VersionBaseSchema({
    stream    : {
        pages           : { type: [pageSchema], default:[] },
        style           : { type: String, trim: true }
    }
});

var Version = mongoose.model('version', versionSchema);
var TypedVersion = Version.discriminator('Test', testSchema);

TypedVersion.create({ language: 'abc', stream: { pages: [{ style: 'abc' }] } }, function(error) {
  assert.ifError(error);
  TypedVersion.findOne({}, { language: 1 }, function(error, version) {
    assert.ifError(error);
    console.log(require('util').inspect(version));
    process.exit(0);
  });
});
```

I get the below output:

```
Mongoose: versions.insert({ language: 'abc', type: 'Test', updated: new Date("Tue, 04 Aug 2015 19:09:49 GMT"), _id: ObjectId("55c10dfd620621a7565d17a1"), stream: { pages: [ { style: 'abc', _id: ObjectId("55c10dfd620621a7565d17a2"), components: [] } ] }, __v: 0 })   
Mongoose: versions.findOne({ type: 'Test' }) { fields: { type: 1, language: 1 } }  
{ _id: 55c1093dcebd26ce4139ab69,
  language: 'abc',
  type: 'Test',
  stream: {} }
```

**Comment by vkarpov15** [comment]
This issue has been stale for a while, re-open if this is still a problem.
